core/src/operations/mod.rs

```diff
@@ -363,10 +363,30 @@ pub enum Operation {
     Pipe,
 
     // ----- cryptographic operations -------------------------------------------------------------
-    /// Applies a permutation of Rescue Prime Optimized to the top 12 elements of the stack. The
-    /// rate part of the sponge is assumed to be on top of the stack, and the capacity is expected
-    /// to be deepest in the stack, starting at stack[8]. For an RPO permutation of [A, B, C] where
-    /// A is the capacity, the stack should look like [C, B, A, ...] from the top.
+    /// Applies a single permutation of Rescue Prime Optimized to the top 12 elements of the stack.
+    ///
+    /// # Usage
+    ///
+    /// This operation performs a single permutation of the Rescue Prime Optimized hash function.
+    /// The top 12 elements of the stack are used as the state for the hash function. The stack
+    /// itself is in reverse order, so the first 8 elements in the range 0..8 contains the two
+    /// elements to be absorbed by the hasher, and the last 4 elements in the range 8..12 contains
+    /// the capacity word of the hash function. The result of the permutation is in the middle
+    /// word, in the range 4..8.
+    ///
+    /// When starting a new hash, the input size must be known. If the input size is even, then the
+    /// capacity word must be initialized to [0,0,0,0], if the size is odd, then the capacity must
+    /// be initialized to [0,0,0,1], meaning the highest element in the VM stack must be set to 1
+    /// (Note: the hash function expects the very first element to be set to 1, here the last
+    /// element is set because the hasher state is in reverse order).
+    ///
+    /// Stack transition:
+    ///
+    /// > [a,b,c,d, e,f,g,h, i,j,k,l] -> [a',b',c',d', e',f',g',h', i',j',k',l']
+    ///
+    /// Used as follows:
+    ///
+    /// > rpo_hash([l,k,j,i h,g,f,e, d,c,b,a]) -> [l',k',j',i', h',g',f',e', d',c',b',a']
```

Comment:
I find this confusing too - maybe using words would help me, and maybe reversing the order so that we see the hasher state first inside of `rpo_hash` and then see the stack transition second would help. For example, for a hasher state of `[A, B, C]`, it appears on the stack as `[C, B, A]`, and I find this less confusing than seeing the stack as `[A, B, C]` and the state as `[C, B, A]` (mainly because this way the first elements of the state are given the lowest alphabetic representation as well, which feels more consistent)

Reply:
replaced with a stack transition based on words.

Unfortunately one issue that I have, is that I never remember where the padding `1` goes. I.e. `push.0.0.0.1` or `push.1.0.0.0`, suggestions on how to documented that are welcome :)